File: aicq-server/src/services/adminService.ts
import bcrypt from 'bcrypt';
import { v4 as uuidv4 } from 'uuid';
import crypto from 'crypto';
import { store } from '../db/memoryStore';
import { config } from '../config';
import { verifyJWT, generateToken } from './accountService';
import type { Account, FriendPermission } from '../models/types';

// ─── Admin Credentials Store ────────────────────────────────────────

interface AdminCredentials {
  username: string;
  passwordHash: string;
  createdAt: number;
}

interface AdminSession {
  id: string;
  token: string;
  createdAt: number;
  expiresAt: number;
}

// In-memory admin state
let adminCredentials: AdminCredentials | null = null;
const adminSessions = new Map<string, AdminSession>();

// Blacklist: accountId -> { reason, addedAt }
const blacklist = new Map<string, { reason: string; addedAt: number }>();

const BCRYPT_ROUNDS = 12;
const ADMIN_SESSION_TTL = 24 * 3600; // 24 hours in seconds

// ─── Admin Auth ────────────────────────────────────────────────────

export function isInitialized(): boolean {
  return adminCredentials !== null;
}

export async function initializeAdmin(username: string, password: string): Promise<{ token: string; expiresIn: number }> {
  if (adminCredentials) {
    throw new Error('管理员已初始化，无法重复设置');
  }

  const passwordHash = await bcrypt.hash(password, BCRYPT_ROUNDS);
  adminCredentials = {
    username,
    passwordHash,
    createdAt: Date.now(),
  };

  console.log(`[admin] Admin initialized: ${username}`);
  return createAdminSession();
}

export async function loginAdmin(username: string, password: string): Promise<{ token: string; expiresIn: number }> {
  if (!adminCredentials) {
    throw new Error('管理员未初始化');
  }

  if (username !== adminCredentials.username) {
    throw new Error('用户名或密码错误');
  }

  const valid = await bcrypt.compare(password, adminCredentials.passwordHash);
  if (!valid) {
    throw new Error('用户名或密码错误');
  }

  console.log(`[admin] Admin logged in: ${username}`);
  return createAdminSession();
}

function createAdminSession(): { token: string; expiresIn: number } {
  const { token, expiresAt } = generateToken(
    { sub: 'admin', type: 'admin' },
    config.jwtSecret + '-admin',
    ADMIN_SESSION_TTL,
  );

  const sessionId = uuidv4();
  adminSessions.set(sessionId, {
    id: sessionId,
    token,
    createdAt: Date.now(),
    expiresAt: expiresAt * 1000,
  });

  return { token, expiresIn: ADMIN_SESSION_TTL };
}

export function verifyAdminSession(token: string): boolean {
  const payload = verifyJWT(token, config.jwtSecret + '-admin');
  if (!payload) return false;
  if (payload.sub !== 'admin' || payload.type !== 'admin') return false;

  // Verify session exists
  for (const [, session] of adminSessions) {
    if (session.token === token && session.expiresAt > Date.now()) {
      return true;
    }
  }

  return false;
}

export function logoutAdmin(token: string): void {
  for (const [id, session] of adminSessions) {
    if (session.token === token) {
      adminSessions.delete(id);
      return;
    }
  }
}

// ─── Dashboard Stats ───────────────────────────────────────────────

export function getDashboardStats(): {
  totalNodes: number;
  totalAccounts: number;
  totalFriendships: number;
  totalGroups: number;
  totalBlacklisted: number;
} {
  let totalFriendships = 0;
  for (const [, node] of store.nodes) {
    totalFriendships += node.friends.size;
  }
  // Each friendship is counted twice (once per node), so divide by 2
  totalFriendships = Math.floor(totalFriendships / 2);

  return {
    totalNodes: store.nodes.size,
    totalAccounts: store.accounts.size,
    totalFriendships,
    totalGroups: store.groups.size,
    totalBlacklisted: blacklist.size,
  };
}

// ─── Node Management ───────────────────────────────────────────────

export interface NodeListItem {
  id: string;
  publicKey: string;
  friendCount: number;
  lastSeen: number;
  isOnline: boolean;
}

export interface NodeListResponse {
  nodes: NodeListItem[];
  total: number;
  page: number;
  pageSize: number;
}

export function getNodeList(params: {
  search?: string;
  page?: number;
  pageSize?: number;
}): NodeListResponse {
  const page = params.page || 1;
  const pageSize = params.pageSize || 20;
  const search = (params.search || '').trim().toLowerCase();

  let nodes = Array.from(store.nodes.values());

  if (search) {
    nodes = nodes.filter(
      (n) =>
        n.id.toLowerCase().includes(search) ||
        n.publicKey.toLowerCase().includes(search),
    );
  }

  // Sort by lastSeen descending (most recent first)
  nodes.sort((a, b) => b.lastSeen - a.lastSeen);

  const total = nodes.length;
  const start = (page - 1) * pageSize;
  const paged = nodes.slice(start, start + pageSize);

  const now = Date.now();

  return {
    nodes: paged.map((n) => ({
      id: n.id,
      publicKey: n.publicKey,
      friendCount: n.friends.size,
      lastSeen: n.lastSeen,
      isOnline: (now - n.lastSeen) < 5 * 60 * 1000, // online if seen within 5 min
    })),
    total,
    page,
    pageSize,
  };
}

export interface FriendInfo {
  nodeId: string;
  lastSeen: number;
  isOnline: boolean;
  permissions: FriendPermission[];
}

export interface NodeDetail {
  id: string;
  publicKey: string;
  lastSeen: number;
  isOnline: boolean;
  socketId: string | null;
  gatewayUrl?: string;
  friends: FriendInfo[];
}

export function getNodeDetail(nodeId: string): NodeDetail | null {
  const node = store.nodes.get(nodeId);
  if (!node) return null;

  const now = Date.now();
  const permissionsMap = store.nodePermissions.get(nodeId);

  const friends: FriendInfo[] = Array.from(node.friends).map((friendId) => {
    const friendNode = store.nodes.get(friendId);
    const perms = permissionsMap?.get(friendId) || ['chat'];
    return {
      nodeId: friendId,
      lastSeen: friendNode?.lastSeen || 0,
      isOnline: friendNode ? (now - friendNode.lastSeen) < 5 * 60 * 1000 : false,
      permissions: perms,
    };
  });

  return {
    id: node.id,
    publicKey: node.publicKey,
    lastSeen: node.lastSeen,
    isOnline: (now - node.lastSeen) < 5 * 60 * 1000,
    socketId: node.socketId,
    gatewayUrl: node.gatewayUrl,
    friends,
  };
}

// ─── Account Management ────────────────────────────────────────────

export interface AccountListItem {
  id: string;
  type: 'human' | 'ai';
  email?: string;
  phone?: string;
  displayName?: string;
  agentName?: string;
  status: 'active' | 'disabled' | 'suspended';
  createdAt: number;
  lastLoginAt: number;
  friendCount: number;
}

export interface AccountListResponse {
  accounts: AccountListItem[];
  total: number;
  page: number;
  pageSize: number;
}

export function getAccountList(params: {
  search?: string;
  page?: number;
  pageSize?: number;
}): AccountListResponse {
  const page = params.page || 1;
  const pageSize = params.pageSize || 20;
  const search = (params.search || '').trim().toLowerCase();

  let accounts = Array.from(store.accounts.values());

  if (search) {
    accounts = accounts.filter(
      (a) =>
        a.id.toLowerCase().includes(search) ||
        (a.email && a.email.toLowerCase().includes(search)) ||
        (a.phone && a.phone.toLowerCase().includes(search)) ||
        (a.displayName && a.displayName.toLowerCase().includes(search)) ||
        (a.agentName && a.agentName.toLowerCase().includes(search)),
    );
  }

  // Sort by createdAt descending
  accounts.sort((a, b) => b.createdAt - a.createdAt);

  const total = accounts.length;
  const start = (page - 1) * pageSize;
  const paged = accounts.slice(start, start + pageSize);

  return {
    accounts: paged.map((a) => ({
      id: a.id,
      type: a.type,
      email: a.email,
      phone: a.phone,
      displayName: a.displayName,
      agentName: a.agentName,
      status: a.status,
      createdAt: a.createdAt,
      lastLoginAt: a.lastLoginAt,
      friendCount: a.friends.length,
    })),
    total,
    page,
    pageSize,
  };
}

export function getAccountDetail(accountId: string): Account | null {
  return store.accounts.get(accountId) || null;
}

export async function createAccount(params: {
  email?: string;
  phone?: string;
  password?: string;
  displayName?: string;
  publicKey: string;
}): Promise<Account> {
  const { email, phone, password, displayName, publicKey } = params;

  if (!email && !phone) {
    throw new Error('必须提供邮箱或手机号');
  }

  if (!publicKey) {
    throw new Error('必须提供公钥');
  }

  // Check for existing accounts with same email/phone
  if (email) {
    const existingEmail = store.emailIndex.get(email);
    if (existingEmail) {
      throw new Error('该邮箱已被注册');
    }
  }

  if (phone) {
    const existingPhone = store.phoneIndex.get(phone);
    if (existingPhone) {
      throw new Error('该手机号已被注册');
    }
  }

  // Check publicKey uniqueness
  const existingKey = store.publicKeyIndex.get(publicKey);
  if (existingKey) {
    throw new Error('该公钥已被使用');
  }

  const accountId = uuidv4();
  let passwordHash: string | undefined;
  if (password) {
    passwordHash = await bcrypt.hash(password, BCRYPT_ROUNDS);
  }

  const account: Account = {
    id: accountId,
    type: 'human',
    email: email || undefined,
    phone: phone || undefined,
    passwordHash,
    displayName: displayName || (email ? email.split('@')[0] : phone || 'User'),
    publicKey,
    createdAt: Date.now(),
    lastLoginAt: Date.now(),
    status: 'active',
    friends: [],
    maxFriends: config.maxFriends,
    visitPermissions: [],
  };

  store.setAccount(account);
  console.log(`[admin] Account created: ${account.id} (${email || phone})`);
  return account;
}

export function deleteAccount(accountId: string): boolean {
  const account = store.accounts.get(accountId);
  if (!account) {
    throw new Error('账号不存在');
  }

  // Remove from indexes
  if (account.email) store.emailIndex.delete(account.email);
  if (account.phone) store.phoneIndex.delete(account.phone);
  if (account.publicKey) store.publicKeyIndex.delete(account.publicKey);

  // Remove account
  store.accounts.delete(accountId);

  // Remove associated sessions
  for (const [sessionId, session] of store.sessions) {
    if (session.accountId === accountId) {
      store.sessions.delete(sessionId);
    }
  }

  // Remove from blacklist
  blacklist.delete(accountId);

  // Remove friend relationships from all nodes
  for (const [, node] of store.nodes) {
    if (node.friends.has(accountId)) {
      node.friends.delete(accountId);
      node.friendCount = node.friends.size;
    }
  }

  // Remove from all groups
  for (const [groupId, group] of store.groups) {
    if (group.members.has(accountId)) {
      group.members.delete(accountId);
    }
  }

  // Remove node record if exists
  store.nodes.delete(accountId);

  console.log(`[admin] Account deleted: ${accountId}`);
  return true;
}

export type AccountUpdateFields = Partial<{
  email: string;
  phone: string;
  displayName: string;
  agentName: string;
  status: 'active' | 'disabled' | 'suspended';
  publicKey: string;
  password: string; // special: will be hashed
  maxFriends: number;
}>;

export async function updateAccount(accountId: string, updates: AccountUpdateFields): Promise<Account> {
  const account = store.accounts.get(accountId);
  if (!account) {
    throw new Error('账号不存在');
  }

  // Handle email change
  if (updates.email !== undefined) {
    if (updates.email && store.emailIndex.get(updates.email) && store.emailIndex.get(updates.email) !== accountId) {
      throw new Error('该邮箱已被其他账号使用');
    }
    if (account.email) store.emailIndex.delete(account.email);
    account.email = updates.email || undefined;
    if (account.email) store.emailIndex.set(account.email, accountId);
  }

  // Handle phone change
  if (updates.phone !== undefined) {
    if (updates.phone && store.phoneIndex.get(updates.phone) && store.phoneIndex.get(updates.phone) !== accountId) {
      throw new Error('该手机号已被其他账号使用');
    }
    if (account.phone) store.phoneIndex.delete(account.phone);
    account.phone = updates.phone || undefined;
    if (account.phone) store.phoneIndex.set(account.phone, accountId);
  }

  // Handle publicKey change
  if (updates.publicKey !== undefined) {
    if (updates.publicKey && store.publicKeyIndex.get(updates.publicKey) && store.publicKeyIndex.get(updates.publicKey) !== accountId) {
      throw new Error('该公钥已被其他账号使用');
    }
    if (account.publicKey) store.publicKeyIndex.delete(account.publicKey);
    account.publicKey = updates.publicKey;
    if (account.publicKey) store.publicKeyIndex.set(account.publicKey, accountId);
  }

  // Handle password change
  if (updates.password !== undefined) {
    if (updates.password) {
      account.passwordHash = await bcrypt.hash(updates.password, BCRYPT_ROUNDS);
    } else {
      account.passwordHash = undefined;
    }
  }

  // Handle simple field updates
  if (updates.displayName !== undefined) account.displayName = updates.displayName;
  if (updates.agentName !== undefined) account.agentName = updates.agentName;
  if (updates.status !== undefined) account.status = updates.status;
  if (updates.maxFriends !== undefined) account.maxFriends = updates.maxFriends;

  store.setAccount(account);
  console.log(`[admin] Account updated: ${accountId}`);
  return account;
}

// ─── Config Management ─────────────────────────────────────────────

export function getConfig(): Record<string, string | number | boolean> {
  return { ...config };
}

export function updateConfig(updates: Record<string, string | number | boolean>): Record<string, string | number | boolean> {
  for (const [key, value] of Object.entries(updates)) {
    if (key === 'jwtSecret') {
      // Don't allow changing jwtSecret via admin API
      console.warn('[admin] Attempted to change jwtSecret via admin API, blocked');
      continue;
    }
    if (key in config) {
      (config as any)[key] = value;
    } else {
      // Allow adding new config fields
      (config as any)[key] = value;
    }
  }

  console.log('[admin] Config updated:', Object.keys(updates).join(', '));
  return { ...config };
}

// ─── Blacklist Management ──────────────────────────────────────────

export interface BlacklistEntry {
  accountId: string;
  reason: string;
  addedAt: number;
}

export function getBlacklist(): BlacklistEntry[] {
  return Array.from(blacklist.entries()).map(([accountId, entry]) => ({
    accountId,
    reason: entry.reason,
    addedAt: entry.addedAt,
  }));
}

export function addToBlacklist(accountId: string, reason: string): BlacklistEntry {
  // Verify account exists
  const account = store.accounts.get(accountId);
  if (!account) {
    throw new Error('账号不存在');
  }

  if (blacklist.has(accountId)) {
    throw new Error('该账号已在黑名单中');
  }

  const entry = {
    reason: reason || '管理员手动添加',
    addedAt: Date.now(),
  };

  blacklist.set(accountId, entry);

  // Disable the account
  account.status = 'suspended';
  store.setAccount(account);

  // Remove all active sessions
  for (const [sessionId, session] of store.sessions) {
    if (session.accountId === accountId) {
      store.sessions.delete(sessionId);
    }
  }

  console.log(`[admin] Account blacklisted: ${accountId} (${reason})`);
  return { accountId, ...entry };
}

export function removeFromBlacklist(accountId: string): boolean {
  if (!blacklist.has(accountId)) {
    throw new Error('该账号不在黑名单中');
  }

  blacklist.delete(accountId);

  // Re-enable the account
  const account = store.accounts.get(accountId);
  if (account && account.status === 'suspended') {
    account.status = 'active';
    store.setAccount(account);
  }

  console.log(`[admin] Account removed from blacklist: ${accountId}`);
  return true;
}

export function isBlacklisted(accountId: string): boolean {
  return blacklist.has(accountId);
}

File: server/services/admin_service.py
"""
AICQ Admin Service
==================
Handles admin initialization, authentication, dashboard statistics,
node/account management, runtime configuration, and blacklist operations.
"""

from __future__ import annotations

import logging
import uuid
from datetime import datetime, timezone, timedelta
from typing import Any, Dict, List, Optional

import bcrypt
import jwt

from db import DatabaseManager, json_serialize, json_list, iso_now
from config import Config

logger = logging.getLogger("aicq.admin_service")


# ─── Password Helpers ───────────────────────────────────────────────────


def _hash_password(password: str) -> str:
    """Hash a plaintext password using bcrypt."""
    salt = bcrypt.gensalt(rounds=12)
    hashed = bcrypt.hashpw(password.encode("utf-8"), salt)
    return hashed.decode("utf-8")


def _verify_password(password: str, password_hash: str) -> bool:
    """Verify a plaintext password against a bcrypt hash."""
    try:
        return bcrypt.checkpw(password.encode("utf-8"), password_hash.encode("utf-8"))
    except Exception:
        return False


# ─── Admin JWT ──────────────────────────────────────────────────────────


def _create_admin_token() -> str:
    """Create an admin JWT token with 24h expiry, signed with JWT_SECRET + '-admin'."""
    now = datetime.now(timezone.utc)
    expiry = now + timedelta(seconds=Config.ADMIN_JWT_EXPIRY)
    payload = {
        "sub": "admin",
        "type": "admin",
        "iat": now,
        "exp": expiry,
    }
    return jwt.encode(payload, Config.JWT_SECRET + "-admin", algorithm="HS256")


def verify_admin_token(token: str) -> Dict[str, Any]:
    """Verify an admin JWT token.

    Returns the decoded payload dict.
    Raises jwt.PyJWTError on invalid or expired tokens.
    """
    payload = jwt.decode(token, Config.JWT_SECRET + "-admin", algorithms=["HS256"])
    if payload.get("type") != "admin":
        raise jwt.InvalidTokenError("Not an admin token")
    return payload


# ─── Admin Initialization ──────────────────────────────────────────────


async def is_initialized(db: DatabaseManager) -> bool:
    """Check if the admin account has been initialized.

    Returns True if an 'admin_password' key exists in admin_config.
    """
    row = await db.fetchone("SELECT value FROM admin_config WHERE key = 'admin_password'")
    return row is not None


async def init_admin(db: DatabaseManager, password: str) -> Dict[str, Any]:
    """First-time admin setup: hash and store the admin password.

    Raises RuntimeError if the admin is already initialized.
    Returns the admin JWT token.
    """
    if await is_initialized(db):
        raise RuntimeError("Admin is already initialized")

    password_hash = _hash_password(password)
    now = iso_now()

    await db.execute(
        "INSERT INTO admin_config (key, value, updated_at) VALUES (?, ?, ?)",
        ("admin_password", password_hash, now),
    )

    token = _create_admin_token()
    logger.info("Admin initialized")
    return {"token": token}


async def login_admin(db: DatabaseManager, password: str) -> Dict[str, Any]:
    """Authenticate admin with password.

    Returns a dict with the admin JWT token.
    Raises ValueError on invalid credentials.
    """
    row = await db.fetchone("SELECT value FROM admin_config WHERE key = 'admin_password'")
    if not row:
        raise ValueError("Admin not initialized")

    if not _verify_password(password, row["value"]):
        raise ValueError("Invalid admin password")

    token = _create_admin_token()
    logger.info("Admin login successful")
    return {"token": token}


# ─── Dashboard Stats ───────────────────────────────────────────────────


async def get_stats(db: DatabaseManager) -> Dict[str, Any]:
    """Return dashboard statistics: total nodes, accounts, friendships, groups, blacklisted count."""
    nodes_count = await db.fetchone("SELECT COUNT(*) as cnt FROM nodes")
    accounts_count = await db.fetchone("SELECT COUNT(*) as cnt FROM accounts")
    friendships_count = await db.fetchone("SELECT COUNT(*) as cnt FROM node_permissions")
    groups_count = await db.fetchone("SELECT COUNT(*) as cnt FROM groups")
    blacklist_count = await db.fetchone("SELECT COUNT(*) as cnt FROM blacklist")

    return {
        "total_nodes": nodes_count["cnt"] if nodes_count else 0,
        "total_accounts": accounts_count["cnt"] if accounts_count else 0,
        "total_friendships": friendships_count["cnt"] if friendships_count else 0,
        "total_groups": groups_count["cnt"] if groups_count else 0,
        "total_blacklisted": blacklist_count["cnt"] if blacklist_count else 0,
    }


# ─── Node Management ───────────────────────────────────────────────────


async def list_nodes(
    db: DatabaseManager,
    page: int = 1,
    per_page: int = 20,
    search: Optional[str] = None,
) -> Dict[str, Any]:
    """Return a paginated list of nodes, optionally filtered by search."""
    offset = (page - 1) * per_page
    params: list[Any] = []

    where_clause = ""
    if search:
        where_clause = "WHERE id LIKE ? OR public_key LIKE ?"
        search_pattern = f"%{search}%"
        params.extend([search_pattern, search_pattern])

    # Count total
    count_row = await db.fetchone(f"SELECT COUNT(*) as cnt FROM nodes {where_clause}", params)
    total = count_row["cnt"] if count_row else 0

    # Fetch page
    query = f"SELECT * FROM nodes {where_clause} ORDER BY last_seen DESC LIMIT ? OFFSET ?"
    params.extend([per_page, offset])
    rows = await db.fetchall(query, params)

    # Deserialize JSON fields
    for row in rows:
        row["friends"] = json_list(row.get("friends"))

    return {
        "nodes": rows,
        "total": total,
        "page": page,
        "per_page": per_page,
        "total_pages": (total + per_page - 1) // per_page,
    }


async def get_node(db: DatabaseManager, node_id: str) -> Optional[Dict[str, Any]]:
    """Get node detail by ID. Returns None if not found."""
    node = await db.fetchone("SELECT * FROM nodes WHERE id = ?", (node_id,))
    if node is None:
        return None
    node["friends"] = json_list(node.get("friends"))
    return node


# ─── Account Management (Admin) ────────────────────────────────────────


async def list_accounts(
    db: DatabaseManager,
    page: int = 1,
    per_page: int = 20,
    search: Optional[str] = None,
) -> Dict[str, Any]:
    """Return a paginated list of accounts, optionally filtered by search."""
    offset = (page - 1) * per_page
    params: list[Any] = []

    where_clause = ""
    if search:
        where_clause = (
            "WHERE id LIKE ? OR email LIKE ? OR display_name LIKE ? "
            "OR agent_name LIKE ? OR public_key LIKE ?"
        )
        search_pattern = f"%{search}%"
        params.extend([search_pattern] * 5)

    count_row = await db.fetchone(f"SELECT COUNT(*) as cnt FROM accounts {where_clause}", params)
    total = count_row["cnt"] if count_row else 0

    query = f"SELECT * FROM accounts {where_clause} ORDER BY created_at DESC LIMIT ? OFFSET ?"
    params.extend([per_page, offset])
    rows = await db.fetchall(query, params)

    # Remove password_hash and deserialize JSON
    for row in rows:
        row.pop("password_hash", None)
        row["friends"] = json_list(row.get("friends"))
        row["visit_permissions"] = json_list(row.get("visit_permissions"))

    return {
        "accounts": rows,
        "total": total,
        "page": page,
        "per_page": per_page,
        "total_pages": (total + per_page - 1) // per_page,
    }


async def create_account(db: DatabaseManager, data: Dict[str, Any]) -> Dict[str, Any]:
    """Admin creates an account.

    Required in data: type, public_key
    For human: email, password, display_name
    For ai: agent_name
    """
    account_type = data.get("type", "human")
    public_key = data.get("public_key")
    if not public_key:
        raise ValueError("public_key is required")

    # Check for duplicate public_key
    existing = await db.fetchone("SELECT id FROM accounts WHERE public_key = ?", (public_key,))
    if existing:
        raise ValueError("Public key already registered")

    account_id = uuid.uuid4().hex
    now = iso_now()

    if account_type == "human":
        email = data.get("email")
        password = data.get("password")
        display_name = data.get("display_name", "")
        phone = data.get("phone")
        password_hash = _hash_password(password) if password else None

        # Check duplicate email
        if email:
            dup = await db.fetchone("SELECT id FROM accounts WHERE email = ?", (email,))
            if dup:
                raise ValueError("Email already registered")

        await db.execute(
            """
            INSERT INTO accounts
                (id, type, email, phone, password_hash, display_name, public_key,
                 created_at, status, friends, visit_permissions)
            VALUES (?, 'human', ?, ?, ?, ?, ?, ?, 'active', '[]', '[]')
            """,
            (account_id, email, phone, password_hash, display_name, public_key, now),
        )
    else:
        agent_name = data.get("agent_name", "")
        await db.execute(
            """
            INSERT INTO accounts
                (id, type, agent_name, public_key, created_at, status, friends, visit_permissions)
            VALUES (?, 'ai', ?, ?, ?, 'active', '[]', '[]')
            """,
            (account_id, agent_name, public_key, now),
        )

    account = await db.fetchone("SELECT * FROM accounts WHERE id = ?", (account_id,))
    account.pop("password_hash", None)
    account["friends"] = json_list(account.get("friends"))
    account["visit_permissions"] = json_list(account.get("visit_permissions"))
    logger.info("Admin created account: %s", account_id)
    return account


async def update_account(
    db: DatabaseManager,
    account_id: str,
    data: Dict[str, Any],
) -> Optional[Dict[str, Any]]:
    """Admin updates an account. Returns the updated account or None."""
    allowed_fields = {
        "display_name", "agent_name", "phone", "email", "public_key",
        "fingerprint", "status", "visit_permissions", "max_friends",
    }

    updates: list[str] = []
    params: list[Any] = []

    for field, value in data.items():
        if field not in allowed_fields:
            continue
        if field in ("visit_permissions",):
            value = json_serialize(value)
        if field == "password":
            value = _hash_password(value)
            field = "password_hash"
        updates.append(f"{field} = ?")
        params.append(value)

    if not updates:
        row = await db.fetchone("SELECT * FROM accounts WHERE id = ?", (account_id,))
        if row is None:
            return None
        row.pop("password_hash", None)
        row["friends"] = json_list(row.get("friends"))
        row["visit_permissions"] = json_list(row.get("visit_permissions"))
        return row

    updates.append("updated_at = ?")
    params.append(iso_now())
    params.append(account_id)

    await db.execute(
        f"UPDATE accounts SET {', '.join(updates)} WHERE id = ?",
        params,
    )

    row = await db.fetchone("SELECT * FROM accounts WHERE id = ?", (account_id,))
    if row is None:
        return None
    row.pop("password_hash", None)
    row["friends"] = json_list(row.get("friends"))
    row["visit_permissions"] = json_list(row.get("visit_permissions"))
    return row


async def delete_account(db: DatabaseManager, account_id: str) -> bool:
    """Admin deletes an account and all related data."""
    # Delegate to account_service which handles cleanup
    from services.account_service import delete_account as _delete_account
    return await _delete_account(db, account_id)


# ─── Runtime Configuration ─────────────────────────────────────────────


async def get_config(db: DatabaseManager) -> Dict[str, Any]:
    """Get all runtime configuration from admin_config table + Config class."""
    rows = await db.fetchall("SELECT key, value FROM admin_config ORDER BY key")
    db_config = {row["key"]: row["value"] for row in rows}

    # Merge with live Config values
    live_config = Config.to_dict(include_secrets=False)

    return {"database": db_config, "runtime": live_config}


async def update_config(db: DatabaseManager, data: Dict[str, Any]) -> Dict[str, Any]:
    """Update runtime configuration.

    Only whitelisted keys are accepted. Updates both the in-memory Config
    and persists to admin_config table.
    """
    result = Config.update_from_dict(data)

    # Persist updated keys to database
    now = iso_now()
    for key in result["updated"]:
        value = getattr(Config, key, None)
        if value is not None:
            await db.execute(
                "INSERT OR REPLACE INTO admin_config (key, value, updated_at) VALUES (?, ?, ?)",
                (key.lower(), str(value), now),
            )

    logger.info("Config update: updated=%s, rejected=%s", result["updated"], result["rejected"])
    return result


# ─── Blacklist Management ──────────────────────────────────────────────


async def get_blacklist(db: DatabaseManager) -> List[Dict[str, Any]]:
    """List all blacklist entries with account info."""
    rows = await db.fetchall(
        """
        SELECT b.id, b.account_id, b.reason, b.created_at,
               a.display_name, a.agent_name, a.email, a.type
        FROM blacklist b
        LEFT JOIN accounts a ON b.account_id = a.id
        ORDER BY b.created_at DESC
        """,
    )
    return rows


async def add_to_blacklist(
    db: DatabaseManager,
    account_id: str,
    reason: Optional[str] = None,
) -> Dict[str, Any]:
    """Add an account to the blacklist and suspend it."""
    # Verify account exists
    account = await db.fetchone("SELECT id FROM accounts WHERE id = ?", (account_id,))
    if not account:
        raise ValueError("Account not found")

    now = iso_now()
    await db.execute(
        "INSERT INTO blacklist (account_id, reason, created_at) VALUES (?, ?, ?)",
        (account_id, reason, now),
    )

    # Suspend the account
    await db.execute(
        "UPDATE accounts SET status = 'suspended', updated_at = ? WHERE id = ?",
        (now, account_id),
    )

    logger.info("Added account %s to blacklist: %s", account_id, reason)
    return {"account_id": account_id, "reason": reason, "created_at": now}


async def remove_from_blacklist(db: DatabaseManager, entry_id: int) -> bool:
    """Remove a blacklist entry by ID and reactivate the account."""
    entry = await db.fetchone("SELECT account_id FROM blacklist WHERE id = ?", (entry_id,))
    if not entry:
        return False

    account_id = entry["account_id"]
    await db.execute("DELETE FROM blacklist WHERE id = ?", (entry_id,))

    # Reactivate the account
    now = iso_now()
    await db.execute(
        "UPDATE accounts SET status = 'active', updated_at = ? WHERE id = ?",
        (now, account_id),
    )

    logger.info("Removed account %s from blacklist (entry %d)", account_id, entry_id)
    return True
